python/ray/data/_internal/logical/rules/inherit_batch_format.py

```diff
@@ -0,0 +1,37 @@
+from collections import deque
+
+from ray.data._internal.logical.interfaces import LogicalOperator, LogicalPlan, Rule
+from ray.data._internal.logical.operators.all_to_all_operator import Sort
+from ray.data._internal.logical.operators.map_operator import MapBatches
+
+
+class InheritBatchFormatRule(Rule):
+    # TODO(xingyu-long): add doc here
+
+    def apply(self, plan: LogicalPlan) -> LogicalPlan:
+        optimized_dag: LogicalOperator = self._apply(plan.dag)
+        new_plan = LogicalPlan(dag=optimized_dag, context=plan.context)
+        return new_plan
+
+    def _apply(self, op: LogicalOperator):
+        # Post-order traversal.
+        nodes = deque()
+        for node in op.post_order_iter():
+            nodes.appendleft(node)
+
+        while len(nodes) > 0:
+            current_op = nodes.pop()
+            upstream_ops = current_op.input_dependencies
+
+            # TODO(xingyu-long): extend this to other LogicalOperator as well
+            # e.g., shuffle, aggregate
+            if isinstance(current_op, Sort):
+                batch_format = None
+                for i in range(len(upstream_ops)):
+                    if isinstance(upstream_ops[i], MapBatches):
+                        if upstream_ops[i]._batch_format is not None:
+                            batch_format = upstream_ops[i]._batch_format
```

Comment:
nit: clean up without indices
```suggestion
                for up_op in upstream_ops:
                    if isinstance(up_op, MapBatches):
                            batch_format = up_op._batch_format
```